elfpy/eth/receipts.py

```diff
@@ -0,0 +1,85 @@
+"""Utilities for handling transaction receipts"""
+from __future__ import annotations
+
+from typing import Any, Sequence
+
+from web3 import Web3
+from web3.contract.contract import Contract, ContractEvent
+from web3.types import ABIEvent, EventData, LogReceipt, TxReceipt
+
+
+def get_transaction_logs(
+    web3: Web3, contract: Contract, tx_receipt: TxReceipt, event_names: Sequence[str] | None = None
+) -> list[dict[str, Any]]:
+    """Decode a transaction receipt.
+
+    Arguments
+    ---------
+    web3 : Web3
+        web3 provider object
+    contract : Contract
+        The contract that emitted the receipt
+    tx_receipt : TxReceipt
+        The emitted receipt after a transaction was completed
+    event_names : Sequence[str] | None
+        If not None, then only return logs with matching event names
+
+
+    Returns
+    -------
+    dict[str, Any]
+        A dictionary containing the decoded logs from the transaction.
+        If event_names is not None, then the returned dict will only
+        include logs that have a corresponding "event" entry.
+    """
+    logs: list[dict[Any, Any]] = []
+    if tx_receipt.get("logs"):
+        for log in tx_receipt["logs"]:
+            event_data, event = get_event_object(web3, contract, log, tx_receipt)
+            if event_data and event:
+                formatted_log = dict(event_data)
+                formatted_log["event"] = event.get("name")
+                if (event_names is not None and formatted_log["event"] in event_names) or (event_names is None):
+                    formatted_log["args"] = dict(event_data["args"])
+                    logs.append(formatted_log)
+    return logs
+
+
+def get_event_object(
+    web3: Web3, contract: Contract, log: LogReceipt, tx_receipt: TxReceipt
+) -> tuple[EventData, ABIEvent] | tuple[None, None]:
+    """Retrieve the event object and anonymous types for a given contract and log.
+
+    Arguments
+    ---------
+    web3 : Web3
+        web3 provider object
+    contract : Contract
+        The contract that emitted the receipt
+    log : LogReceipt
+        A TypedDict parsed out of the transaction receipt
+    tx_receipt: TxReceipt
+        The emitted receipt after a transaction was completed
+
+    Returns
+    -------
+    tuple[EventData, ABIEvent] | tuple[None, None]
+        If faile, return (None, None). Otherwise, return the decoded event information as (data, abi).
+    """
+    abi_events = [abi for abi in contract.abi if abi["type"] == "event"]  # type: ignore
```

Comment:
get rid of type ignore:
 
 ```python
from web3.types import ABIEvent
    
abi_events = [ABIEvent(anonymous=abi.anonymous...) for abi in contract.abi if abi["type"] == "event"]  # type: ignore
```

might be a better way to populate ABIEvent with some splatting, not sure of syntax.

Reply:
ok yeah I did this too in elfpy/eth/errors.py:
```python
    errors = [
        ABIError(name=err.get("name"), inputs=err.get("inputs"), type="error")  # type: ignore
        for err in abi
        if err.get("type") == "error"
    ]
```

Reply:
oh i noticed I still have type ignore here, that's because there's no ABIError in web3.py yet.  You shouldn't need it.  I should add a comment about that on mine.